refactor(filter-stream): Refactor & optimize FilterStream

jhermsmeier opened the pull request:
**Changes:**

- Add FilterStream benchmark
- Add `debug` output to FilterStream
- Generate pre-processed ranges to simplify range bounds checking
- Optimize `_transformBlock()` to only slice & dice chunks that don't adhere to the block size, also improving performance considerably by causing fewer operations on buffers
- Fix `_flush()` to emit correctly transformed blocks
- Update emitted blocks check in tests

Change-Type: minor

zvin commented on lib/filter-stream.js:
```diff
@@ -83,79 +79,138 @@ function FilterStream( blockMap, options ) {
-    if( this.currentRange == null ) return
+    // debug( 'range-in-chunk', rangeStart, '-', this.currentRange.end )
+    // debug( 'range-in-chunk:start', rangeStart, '>=', this.position )
+    // debug( 'range-in-chunk:end', rangeStart, '<', this.position + chunk.length )
```
Is there a reason to keep these 3 commented lines?

jhermsmeier replied:
No, those are leftovers – removed now.

zvin commented on test/filter-stream.js:
```diff
@@ -15,14 +15,15 @@ describe( 'BlockMap.FilterStream', function() {
+        blockCount += block.length / blockMap.blockSize
+        assert.equal( block.length % blockMap.blockSize, 0, 'Invalid block size: ' + block.length )
         assert.ok( block.address != null, 'block address missing' )
+        assert.equal( block.address, Math.abs( block.address ), 'block address must be an integer' )
```
NIT: you can use `Number.isInteger(block.address)` for that

jhermsmeier replied:
Done!

zvin commented on lib/filter-stream.js:
```diff
@@ -168,39 +223,52 @@ function FilterStream( blockMap, options ) {
-      this._pushBlock( buffer.slice( offset, offset + this.blockSize ) )
-      offset = offset + this.blockSize
+    if( !this._rangeInChunk( chunk ) ) {
+      debug( 'chunk:ignore' )
```
I'm not sure I get the purpose of FilterStream.
As I understand it, it strips out the parts of the input stream that are not in the blockmap. What would be the use case for this?

jhermsmeier replied:
Yeah, that's the point – you pipe a readable stream into the BlockMap.FilterStream and it it'll filter out unmapped blocks.

One of the main use cases for this is when images from sources that cannot be randomly accessed – i.e. an image being decompressed or extracted from an archive, or an HTTP stream – which is the case in Etcher, as the only way to flash blockmapped images at the moment is to have them in a .zip with the metadata folder alongside it.

zvin replied:
The thing I don't understand is that, if we skip discarded blocks, how Etcher (or anything else using FilterStream) will make to write the non discarded blocks in the correct place?
Shouldn't FilterStream be replacing the discarded blocks with zero (or random) -filled buffers that are the same size?
Or is there some mechanism to skip to the correct position in the output?

jhermsmeier replied:
The emitted buffers have `position` and `address` properties added to them (see [line 175](https://github.com/resin-io-modules/blockmap/pull/27/files#diff-7c60d2150dd8ff0dfcf7b7edb6417d37R173)) – this is going to change soon to a saner way (emitting objects of the form `{ position, address, data }` to utilize hidden classes and make sure that consuming streams don't accidentally write blocks in sequence), but for backwards compatibility this PR doesn't introduce that yet.

zvin replied:
Thanks for the clarification.

jviotti commented on lib/filter-stream.js:
```diff
@@ -48,22 +47,19 @@ function FilterStream( blockMap, options ) {
-   * @private
-   */
-  this._buffers = []
+  this.highWaterMark = 64 * 1024
```
Just out of curiosity: is there any particular reason behind this high water mark number? If so, should we explain it in a comment?

jviotti replied:
Also, do you think it'd be beneficial to allow the client pass a custom one?

jhermsmeier replied:
64K is Node's default highWaterMark for `fs` streams, but other streams come with 16K default highWaterMark – so this is set to the 64K to match fs write ops (it's the speediest this way) – comment added.

jviotti commented on lib/filter-stream.js:
```diff
@@ -83,79 +79,134 @@ function FilterStream( blockMap, options ) {
   constructor: FilterStream,
 
+  get highWaterMark() { return this._writableState.highWaterMark },
+  set highWaterMark( value ) { this._writableState.highWaterMark = value },
```
Ah, ignore my previous comment :P

jviotti commented on lib/filter-stream.js:
```diff
@@ -83,79 +79,134 @@ function FilterStream( blockMap, options ) {
-    return this.currentRange &&
-      blockAddress >= this.currentRange.start &&
-      blockAddress <= this.currentRange.end
+  _prepareRanges() {
```
Maybe we should find a better name for this function? Something like `getByteRangesFromBlockMap`, as it currently doesn't reflect very well the description you put in JSDoc?

jhermsmeier replied:
I agree `prepareRanges()` is a bit vague here – changed!

jhermsmeier commented:
I have now also added comments throughout the methods, and added some missing doc blocks, to hopefully make it easier / quicker to grasp what's going on in there.
